### PR TITLE
[SelectionDAG]: Have isKnownNeverZero treat SRL like division if all else fails

### PR DESCRIPTION
**Pull request by AZero13**

### CODE REVIEW

File: llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp [hunk 1]
```diff
@@ -5443,10 +5443,17 @@ bool SelectionDAG::isKnownNeverZero(SDValue Op, unsigned Depth) const {
...
+    const KnownBits Shift = computeKnownBits(Op.getOperand(1), Depth + 1);
+    APInt MaxCnt = Shift.getMaxValue();
     if (MaxCnt.ult(ValKnown.getBitWidth()) &&
         !ValKnown.One.lshr(MaxCnt).isZero())
       return true;
+    // We try to see if we can turn it into a division
+    const KnownBits One =
+        KnownBits::makeConstant(APInt(ValKnown.getBitWidth(), 1));
+
+    if (KnownBits::uge(ValKnown, KnownBits::lshr(One, Shift)))
```
Comment (topperc) [inline llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp:5455]: Doesn't SRA reach here too? Is this also correct for SRA?

Comment (AtariDreams) [reply llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp:5455]: Yes because SRA only matters if the value is negative, which returns true anyway.

Comment (topperc) [reply llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp:5455]: What if the sign bit is unknown? Then it might be negative.

Comment (topperc) [inline llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp:5455]: Is this supposed to be KnownBits::shl? KnownBits::lshr of 1 will either be 0 or 1 depending on whether the shift is non-zero.

File: llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp [hunk 2]
```diff
@@ -5443,10 +5443,19 @@ bool SelectionDAG::isKnownNeverZero(SDValue Op, unsigned Depth) const {
     if (ValKnown.isNegative())
       return true;
     // If max shift cnt of known ones is non-zero, result is non-zero.
-    APInt MaxCnt = computeKnownBits(Op.getOperand(1), Depth + 1).getMaxValue();
+    const KnownBits Shift = computeKnownBits(Op.getOperand(1), Depth + 1);
+    APInt MaxCnt = Shift.getMaxValue();
     if (MaxCnt.ult(ValKnown.getBitWidth()) &&
         !ValKnown.One.lshr(MaxCnt).isZero())
       return true;
```
Comment (RKSimon) [inline llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp:5450]: Split this so the code below can all assume MaxCnt < BW:
```c
if (MaxCnt.uge(ValKnown.getBitWidth())
  return false;
if (!ValKnown.One.lshr(MaxCnt).isZero())
  return true;
```

**Review by RKSimon — COMMENTED** [review]
Please can you try to more closely match the ValueTracking.cpp implementation which tries to do the same thing - I don't think you need to bother adding a isNonZeroShift helper but you should be trying to use the same methodology.

**Comment by jayfoad** [comment]
What cases does this handle that were not handled already? Intuitively it seems like `KnownBits::shl(One, Shift)` will not contain any useful information unless we already knew the exact value of `Shift`.

**Pull request closed without merging**